src/common/types.ts

```diff
@@ -160,12 +161,246 @@ export type CSVRecord<Header extends ReadonlyArray<string>> = Record<
   string
 >;
 
+type Concat<T extends any[]> = T extends [infer F, ...infer R]
+  ? R extends any[]
+    ? F & Concat<R>
+    : F
+  : // biome-ignore lint/complexity/noBannedTypes: <explanation>
+    {};
+
+type Replace<
+  Target extends string,
+  From extends string,
+  To extends string,
+> = Target extends `${infer A}${From}${infer B}`
+  ? Replace<`${A}${To}${B}`, From, To>
+  : Target;
+
+type Split<
+  Char extends string,
+  Delimiter extends string = typeof COMMA,
+> = Char extends `${infer F}${Delimiter}${infer R}`
+  ? [F, ...Split<R, Delimiter>]
+  : [Char];
+
+type Join<
+  Chars extends ReadonlyArray<string | number | boolean | bigint>,
+  Delimiter extends string = typeof COMMA,
+  Quotation extends string = typeof DOUBLE_QUOTE,
+> = Chars extends readonly [infer F, ...infer R]
+  ? F extends string
+    ? R extends string[]
+      ? `${F extends `${string}${Newline | Delimiter}${string}`
+          ? `${Quotation}${F}${Quotation}`
+          : F}${R extends [] ? "" : Delimiter}${Join<R, Delimiter, Quotation>}`
+      : string
+    : string
+  : "";
+
+type Newline = typeof CR | typeof CRLF | typeof LF;
+type DummyNewline<NL extends Newline> =
+  `web-csv-toolbox.DummyNewline-${NL extends typeof CRLF
+    ? "crlf"
+    : NL extends typeof CR
+      ? "cr"
+      : "lf"}`;
+type DummyDelimiter = "web-csv-toolbox.DummyDelimiter";
+
+type SplitNewline<T extends string> =
```

Comment:
While reviewing the TypeScript types for handling newline splitting within CSV data, I've observed an issue that seems to be related to the evaluation order of types. Specifically, this issue affects how string literals containing newline characters are processed, leading to inconsistent behavior across similar test cases.

Here's a brief overview of the problem:

- **Case 1** successfully matches the expected type because the newline (`\n`) within the double quotes is not treated as a separator, which is correct. However, it seems to handle \r\n outside of the quotes correctly.
   ```ts
    // case 1
    expectTypeOf<SplitNewline<'"a\n"\r\nb\r\nc'>>().toMatchTypeOf<
        ['"a\n"', "b", "c"]
    >();
    ```
- **Case 2** fails presumably because the parser incorrectly splits on the `\r\n` sequence inside the quotes, which should ideally be ignored as it is part of the quoted string.

   ```ts
    // case 2
    expectTypeOf<SplitNewline<'"a\r\n"\r\nb\r\nc'>>().toMatchTypeOf<
         ['"a\r\n"', "b", "c"]
    >();
    ```
This suggests that our current method of handling line breaks inside and outside of quoted strings may need refinement to ensure that it reliably distinguishes between these contexts. The difference in behavior between these cases highlights a potential improvement area in our type definitions, possibly requiring a more sophisticated pattern or mechanism to accurately identify and ignore delimiters within quoted segments.